# 1. Desenvolvimento do Modelo

O notebook tem o objetivo de aplicar o modelo de clusterização.

In [1]:
# Importando os pacotes necessários para o Notebook
import pandas as pd
# Normalização
from sklearn.preprocessing import Normalizer
import numpy as np

In [2]:
# Importando a Base de Dados
df_municipios = pd.read_csv('../data/processed/base_tratada.csv', 
                        decimal = ".", 
                        sep = ";")
df_municipios = df_municipios.dropna()
df_municipios.head()

Código        Município  Área (km²)  Densidade demográfica, 2000  \
0  150010  Abaetetuba (PA)      1613.9                         73.8   
1  290070  Alagoinhas (BA)       736.8                        175.9   
2  430060    Alvorada (RS)        72.9                       2516.1   
3  350160   Americana (SP)       133.9                       1359.9   
4  150080  Ananindeua (PA)       191.4                       2053.0   

   Distância à capital (km)  Esperança de vida ao nascer, 2000  \
0                 51.184552                              70.77   
1                 93.320226                              65.34   
2                 14.838698                              69.98   
3                114.560278                              72.46   
4                 17.749114                              72.19   

   Taxa de fecundidade total, 2000  Renda per Capita, 2000  \
0                             3.15                  102.14   
1                             2.95                  181.26   
2                             2.80                  214.75   
3                             1.96                  473.23   
4                             2.35                  187.79   

   Índice de Gini, 2000  Intensidade da indigência, 2000  ...  \
0                  0.59                            44.61  ...   
1                  0.60                            54.63  ...   
2                  0.45                            56.91  ...   
3                  0.50                            77.52  ...   
4                  0.54                            50.21  ...   

   Índice de Desenvolvimento Humano Municipal, 2000  \
0                                             0.706   
1                                             0.729   
2                                             0.768   
3                                             0.840   
4                                             0.782   

   Taxa bruta de freqüência à escola, 2000  Taxa de alfabetização, 2000  \
0                                    81.76                        80.74   
1                                    90.61                        85.71   
2                                    77.40                        94.01   
3                                    87.15                        95.62   
4                                    84.70                        94.57   

   Média de anos de estudo das pessoas de 25 anos ou mais de idade, 2000  \
0                                               4.28                       
1                                               5.82                       
2                                               5.86                       
3                                               6.93                       
4                                               7.01                       

   Populacao  QTD_TOTAL_CNPJ_SUPERMERCADO  Variacao_Populacao_1991_a_2000  \
0     119152                        254.0                        0.191651   
1     130095                        287.0                        0.112931   
2     183968                        636.0                        0.295130   
3     182593                        348.0                        0.186902   
4     393569                        640.0                        0.610071   

   Percentual_Populacao_65_anos  Percentual_Populacao_Urbana  \
0                      0.039345                     0.594560   
1                      0.061563                     0.864291   
2                      0.039371                     0.996722   
3                      0.069576                     0.997623   
4                      0.028041                     0.997607   

   Populacao por Supermercados  
0                   469.102362  
1                   453.292683  
2                   289.257862  
3                   524.692529  
4                   614.951563  

[5 rows x 21 columns]

In [3]:
df_municipios.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 223 entries, 0 to 222
Data columns (total 21 columns):
 #   Column                                                                 Non-Null Count  Dtype  
---  ------                                                                 --------------  -----  
 0   Código                                                                 223 non-null    int64  
 1   Município                                                              223 non-null    object 
 2   Área (km²)                                                             223 non-null    float64
 3   Densidade demográfica, 2000                                            223 non-null    float64
 4   Distância à capital (km)                                               223 non-null    float64
 5   Esperança de vida ao nascer, 2000                                      223 non-null    float64
 6   Taxa de fecundidade total, 2000                                        223 non-null    flo

In [4]:
# Removendo dados para modelo
# df_modelo = df_municipios
df_modelo = df_municipios.drop(columns = ["Área (km²)", \
                                          'Distância à capital (km)', \
                                          'Intensidade da indigência, 2000', \
                                          'Esperança de vida ao nascer, 2000', \
                                          'Taxa bruta de freqüência à escola, 2000', \
                                          'Média de anos de estudo das pessoas de 25 anos ou mais de idade, 2000', \
                                          'Populacao', \
                                          'Taxa de alfabetização, 2000', \
                                          'QTD_TOTAL_CNPJ_SUPERMERCADO', \
                                          'Percentual_Populacao_65_anos', \
                                          'Índice de Gini, 2000', \
                                          'Taxa de fecundidade total, 2000'])
df_modelo.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 223 entries, 0 to 222
Data columns (total 9 columns):
 #   Column                                            Non-Null Count  Dtype  
---  ------                                            --------------  -----  
 0   Código                                            223 non-null    int64  
 1   Município                                         223 non-null    object 
 2   Densidade demográfica, 2000                       223 non-null    float64
 3   Renda per Capita, 2000                            223 non-null    float64
 4   Intensidade da pobreza, 2000                      223 non-null    float64
 5   Índice de Desenvolvimento Humano Municipal, 2000  223 non-null    float64
 6   Variacao_Populacao_1991_a_2000                    223 non-null    float64
 7   Percentual_Populacao_Urbana                       223 non-null    float64
 8   Populacao por Supermercados                       223 non-null    float64
dtypes: float64(7), int64(1

## 2. Normalização dos Dados
Visto que as variáveis tem escalas diferentes é necessário normalizar para que todos tenham a mesma escala.

In [5]:
norm_df = df_modelo.drop(columns = ['Código', 'Município'])
norm_df = norm_df.dropna()
norm_df = Normalizer().fit_transform(norm_df)
norm_df

array([[1.51103290e-01, 2.09128591e-01, 1.04502871e-01, ...,
        3.92399849e-04, 1.21734357e-03, 9.60473036e-01],
       [3.37393406e-01, 3.47674411e-01, 9.65570442e-02, ...,
        2.16613422e-04, 1.65779560e-03, 8.69459706e-01],
       [9.89778017e-01, 8.44778940e-02, 1.59711409e-02, ...,
        1.16097506e-04, 3.92088462e-04, 1.13787637e-01],
       ...,
       [2.35912143e-01, 5.92382330e-01, 1.32295829e-01, ...,
        4.80406024e-04, 2.48412795e-03, 7.58891203e-01],
       [5.77603487e-01, 2.12880145e-01, 8.00214637e-02, ...,
        1.65827463e-04, 1.39079322e-03, 7.83995962e-01],
       [9.34597135e-01, 2.44955348e-01, 3.06466811e-02, ...,
        6.37485311e-05, 7.03356256e-04, 2.56095734e-01]])

In [6]:
transform = df_modelo.drop(columns = ['Código', 'Município'])
transform = transform.dropna()
transform = Normalizer().fit(transform)
# transform.transform(teste)

## 3. Aplicando a Clusterização
### 3.1. k-means

In [7]:
from sklearn.cluster import KMeans
agrupador = KMeans(n_clusters = 10)
agrupador.fit(norm_df)
labels = agrupador.labels_

In [8]:
from sklearn.cluster import KMeans
from sklearn import metrics

# Calculando o Coeficiente de Silhouette
def teste_kmeans(df, n_clusters):
    np.random.seed(75243)
    agrupador = KMeans(n_clusters = n_clusters)
    agrupador.fit(df)
    labels = agrupador.labels_
    silhouette = metrics.silhouette_score(df, labels, metric="euclidean")
    model_score = {'n_clusters': n_clusters,
                   'silhouette': silhouette}
    return model_score

In [9]:
for i in range(2, 10):
    print(teste_kmeans(norm_df, i))

{'n_clusters': 2, 'silhouette': 0.5703263316079199}
{'n_clusters': 3, 'silhouette': 0.4581345187866969}
{'n_clusters': 4, 'silhouette': 0.4860758831282236}
{'n_clusters': 5, 'silhouette': 0.4534282706990402}
{'n_clusters': 6, 'silhouette': 0.4359649407502018}
{'n_clusters': 7, 'silhouette': 0.4331226776834987}
{'n_clusters': 8, 'silhouette': 0.3934433301981291}
{'n_clusters': 9, 'silhouette': 0.3841817982502604}


In [10]:
np.random.seed(75243)
agrupador = KMeans(n_clusters = 4)
agrupador.fit(norm_df)
labels_kmeans = agrupador.labels_
silhouette = metrics.silhouette_score(norm_df, labels, metric="euclidean")

In [11]:
agrupador.predict(norm_df[181:182])

array([3], dtype=int32)

In [22]:
import plotly.graph_objects as go
norm_df = pd.DataFrame(norm_df, columns = df_modelo.drop(columns = ['Código', 'Município']).columns)
fig = go.Figure()
fig.add_trace(go.Scatter(x = norm_df['Populacao por Supermercados'], 
                         y = norm_df['Renda per Capita, 2000'],
                         mode = 'markers',
                         marker = dict(color = labels_kmeans.astype(np.float)),
                         text = labels_kmeans))
fig.update_layout(
    xaxis_title="População por Supermercados",
    yaxis_title="Renda per Capita",
    font=dict(
        
        size=18,
        color="#7f7f7f"
    )
)

fig.show()

<ipython-input-22-a3b13df00e62>:7: DeprecationWarning:

`np.float` is a deprecated alias for the builtin `float`. To silence this warning, use `float` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.float64` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations



In [13]:
norm_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 223 entries, 0 to 222
Data columns (total 7 columns):
 #   Column                                            Non-Null Count  Dtype  
---  ------                                            --------------  -----  
 0   Densidade demográfica, 2000                       223 non-null    float64
 1   Renda per Capita, 2000                            223 non-null    float64
 2   Intensidade da pobreza, 2000                      223 non-null    float64
 3   Índice de Desenvolvimento Humano Municipal, 2000  223 non-null    float64
 4   Variacao_Populacao_1991_a_2000                    223 non-null    float64
 5   Percentual_Populacao_Urbana                       223 non-null    float64
 6   Populacao por Supermercados                       223 non-null    float64
dtypes: float64(7)
memory usage: 12.3 KB


In [23]:
fig = go.Figure()
fig.add_trace(go.Scatter(x = norm_df['Populacao por Supermercados'], 
                         y = norm_df['Densidade demográfica, 2000'],
                         mode = 'markers',
                         marker = dict(color = labels_kmeans.astype(np.float)),
                         text = labels_kmeans))
fig.update_layout(
    xaxis_title="População por Supermercados",
    yaxis_title="Densidade demográfica",
    font=dict(
        
        size=18,
        color="#7f7f7f"
    )
)
fig.show()

<ipython-input-23-bb1d1e0ec6d8>:5: DeprecationWarning:

`np.float` is a deprecated alias for the builtin `float`. To silence this warning, use `float` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.float64` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations



In [15]:
df_municipios["cluster"] = labels_kmeans
description = df_municipios.groupby("cluster")
n_clients = description.size()
description = description.mean()
description['n_clients'] = n_clients
description.drop(columns = ['Código']).T

cluster                                                         0  \
Área (km²)                                             253.678082   
Densidade demográfica, 2000                           3330.330137   
Distância à capital (km)                                31.455219   
Esperança de vida ao nascer, 2000                       70.318082   
Taxa de fecundidade total, 2000                          2.176164   
Renda per Capita, 2000                                 349.061096   
Índice de Gini, 2000                                     0.550959   
Intensidade da indigência, 2000                         61.523151   
Intensidade da pobreza, 2000                            46.157397   
Índice de Desenvolvimento Humano Municipal, 2000         0.795863   
Taxa bruta de freqüência à escola, 2000                 84.739315   
Taxa de alfabetização, 2000                             92.335479   
Média de anos de estudo das pessoas de 25 anos ...       6.743836   
Populacao                                           721475.013699   
QTD_TOTAL_CNPJ_SUPERMERCADO                           1668.972603   
Variacao_Populacao_1991_a_2000                           0.269508   
Percentual_Populacao_65_anos                             0.049706   
Percentual_Populacao_Urbana                              0.989257   
Populacao por Supermercados                            502.326456   
n_clients                                               73.000000   

cluster                                                         1  \
Área (km²)                                             492.929412   
Densidade demográfica, 2000                            621.384314   
Distância à capital (km)                               111.297856   
Esperança de vida ao nascer, 2000                       70.414510   
Taxa de fecundidade total, 2000                          2.270392   
Renda per Capita, 2000                                 318.289216   
Índice de Gini, 2000                                     0.551373   
Intensidade da indigência, 2000                         59.336667   
Intensidade da pobreza, 2000                            44.404510   
Índice de Desenvolvimento Humano Municipal, 2000         0.787784   
Taxa bruta de freqüência à escola, 2000                 83.480392   
Taxa de alfabetização, 2000                             91.102549   
Média de anos de estudo das pessoas de 25 anos ...       6.360588   
Populacao                                           237905.470588   
QTD_TOTAL_CNPJ_SUPERMERCADO                            563.039216   
Variacao_Populacao_1991_a_2000                           0.347676   
Percentual_Populacao_65_anos                             0.049851   
Percentual_Populacao_Urbana                              0.942671   
Populacao por Supermercados                            576.367958   
n_clients                                               51.000000   

cluster                                                         2  \
Área (km²)                                            3297.048571   
Densidade demográfica, 2000                            131.557143   
Distância à capital (km)                               208.612197   
Esperança de vida ao nascer, 2000                       70.695000   
Taxa de fecundidade total, 2000                          2.248286   
Renda per Capita, 2000                                 347.123143   
Índice de Gini, 2000                                     0.580000   
Intensidade da indigência, 2000                         53.534571   
Intensidade da pobreza, 2000                            40.553429   
Índice de Desenvolvimento Humano Municipal, 2000         0.800329   
Taxa bruta de freqüência à escola, 2000                 84.660857   
Taxa de alfabetização, 2000                             91.969714   
Média de anos de estudo das pessoas de 25 anos ...       6.554714   
Populacao                                           244259.542857   
QTD_TOTAL_CNPJ_SUPERMERCADO                            857.428571

Grupos:
- Grupo formado com 73 cidades, composto por capitais e cidades metropolitantas mais populosas, com média da população de 721 mil habitantes e densidade demográfica de 3.330. Cada supermercado nesta região atende em média 502 pessoas, por mais que possuam o maior número de CNPJs de supermercados ainda tem espaço para competitividade. Exemplo: Rio de Janeiro (RJ), São Paulo (SP).


- 51 cidades, com média de 110 km de distância da capital. Cresceram em média de 34% entre 1991 e 2000. Cada Supermercado desta região atende em média 576 pessoas, maior oportunidade de entrada para novos players. Exemplos: São José dos Campos (SP), Joinville (SC).


- Grupo com 29 cidades, com média de 229km da capital, renda per capita de R$ 200,34. Grupo com menor crescimento entre 1991 e 2000 (22,52%) e maior percentual da população maior que 65 anos (5,13%). Exemplo: Santarém (PA), Caruaru (PE).


- 70 cidades, composto pelas cidades de maior território, com menor densidade demográfica (131). Maior renda per capita de todos os clusters R$ 357. Apresenta o maior IDH de todos os grupos. Cada supermercado atende em média 321 pessoas, este cluster apresenta o maior número de competitividade dentre todos. Exemplo: Brasília (DF), Manaus (AM).




In [28]:
df_municipios.query('cluster == 3')[['Município', 'Densidade demográfica, 2000', 'Populacao', 'Populacao por Supermercados']].sort_values(by='Populacao', ascending = False)

Município  Densidade demográfica, 2000  Populacao  \
181                 Santarém (PA)                         10.8     262538   
47                   Caruaru (PE)                        271.8     253634   
51                   Caucaia (CE)                        209.3     250479   
89                    Ilhéus (BA)                        120.1     222127   
150                Petrolina (PE)                         45.9     218538   
131                  Mossoró (RN)                        100.4     213841   
127                  Marília (SP)                        168.1     197342   
200              Sete Lagoas (MG)                        341.9     184871   
67               Divinópolis (MG)                        258.4     183962   
109                 Juazeiro (BA)                         27.1     174567   
38                  Camaçari (BA)                        211.3     161727   
201                   Sobral (CE)                         72.9     155276   
34   Cabo de Santo Agostinho (PE)                        340.8     152977   
105                   Jequié (BA)                         48.3     147202   
52                    Caxias (MA)                         26.3     139756   
49                 Castanhal (PA)                        130.6     134496   
21                 Barreiras (BA)                         11.0     131849   
1                 Alagoinhas (BA)                        175.9     130095   
211                    Timon (MA)                         74.7     129692   
162                Queimados (RJ)                       1560.0     121993   
6             Angra dos Reis (RJ)                        146.0     119247   
0                 Abaetetuba (PA)                         73.8     119152   
180               Santa Rita (PB)                        150.9     115844   
56                  Colatina (ES)                         62.9     112711   
104                      Jaú (SP)                        162.0     112104   
55                      Codó (MA)                         26.1     111146   
103           Jaraguá do Sul (SC)                        200.7     108489   
207      Teixeira de Freitas (BA)                         92.7     107486   
61                     Crato (CE)                         93.4     104646   

     Populacao por Supermercados  
181                   421.409310  
47                    592.602804  
51                    525.113208  
89                    458.940083  
150                   386.792920  
131                   528.002469  
127                   704.792857  
200                   644.149826  
67                    619.400673  
109                   354.091278  
38                    502.257764  
201                   446.195402  
34                    616.842742  
105                   415.824859  
52                    443.669841  
49                    533.714286  
21                    459.404181  
1                     453.292683  
211                   749.664740  
162                  5304.043478  
6                     520.729258  
0                     469.102362  
180                   449.007752  
56                    440.277344  
104                   640.594286  
55                    424.221374  
103                   809.619403  
207                   707.144737  
61                    399.412214

## 4. Exportando a base Modificada
Exportando a base para o notebook para Elaboração do Modelo de Clusterização

In [29]:
df_municipios.to_csv('../data/processed/base_clusterizacao.csv',
                   sep=";",
                   index=False)